language/move-model/src/ty.rs

```diff
@@ -749,16 +749,28 @@ impl TypeUnificationAdapter {
     fn new<'a, I>(
         lhs_types: I,
         rhs_types: I,
-        treat_lhs_type_param_as_var: bool,
-        treat_rhs_type_param_as_var: bool,
+        treat_lhs_type_param_as_var_after_index: Option<TypeParameterIndex>,
```

Comment:
I am still experimenting on this progressive instantiation idea (instead of the cartesian product, which I now feel unsafe about it). I hope that at the end of the experiment we don't need to go through this complication. But at this point, I am not sure.